### PR TITLE
Ef database support implementation examples

### PR DESCRIPTION
**Pull request by addisonbeck**
### Related Work
- [Asana ticket](https://app.asana.com/0/1199950673439194/1153357960682290/f)
- Github issue #453

### Code Changes
1. Created an enum for supported database providers
2. I removed several `JsonDocument` properties from EF models. I think these existed to use the appropriate data types on  per provider level, but data wasn't saving or returning properly. Instead to handle this I:
3. added column type information to `DatabaseContext.cs` where applicable.
4. Added data annotations to several models to enforce things like `MaxLength` in any databases
5. Changed the way `SetNewId()` works on auto generated ids. Right now we do nothing, but EF will use a non-0 number as an Id instead of auto incrementing if one is supplied, so I am manually setting it to 0 in these cases to make sure the values auto generate properly.
6. Added a `ClearChangeTracking()` method to BaseEntityFrameworkRepository to use during unit testing to make sure we are getting data from the DB when that is needed.
7. Organized `OnModelCreating()`, though this may continue to change as development continues.
8. Implemented anything that was missing from the User and Organization EF Repos
9. Implemented the SsoConfig and SsoUser EF repos
10. Unit tested all available EF Repos to make sure they work and that data matches between provider types and Dapper
11. Changed the way we delete data in `Repository.DeleteAsync` for EF repos. Setting the entity state to Deleted and saving changes doesn't account for things like foreign keys, while `.Remove` does.
12. I discovered a few things that aren't working in the Dapper repos. Some methods that appear to be available in the code don't have backing stored procs and views. None of these broken methods are being used, but I was in the area and needed them working to compare implementations, so I added any that I saw. This includes: `SsoConfig_DeleteById`, `SsoUser_ReadyById`, `SsoUser_DeleteById`, and `SsoUserView`.
13. Changed the base Create methods to return the object that was sent with the ID that is saved. This was needed to write unit tests showing data saving appropriately. This involved adding OUTPUT to the Id of any Guid-Id'd stored procs.
14. Added Moq to Core.Test. This was needed for mocking out the ServiceScope stuff used for DbContext lifetime. 

### Notes
- All unit tests are being skipped in CI for the time being. We will need to set up database instances to run them against and have a migration ready before they will work in CI.

### CODE REVIEW

File: src/Core/Enums/SupportedDatabaseProviders.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+namespace Bit.Core.Enums
+{
+    public enum SupportedDatabaseProviders 
+    {
+        MsSql,
```
Comment (kspearrin) [inline src/Core/Enums/SupportedDatabaseProviders.cs:5]: Maybe call it SqlServer, which matches the repo namespace? To me, this always ends up looking too close to MySql when reading

File: src/Core/Repositories/EntityFramework/OrganizationRepository.cs [hunk 2]
```diff
@@ -41,22 +41,33 @@ public async Task<ICollection<TableModel.Organization>> GetManyByEnabledAsync()
...
+                        .Select(ou => ou.Organization))
+                    .ToListAsync();
+                return Mapper.Map<List<TableModel.Organization>>(organizations);
+            }
         }
 
-        public async Task<ICollection<TableModel.Organization>> SearchAsync(string name, string userEmail, bool? paid,
-            int skip, int take)
+        public async Task<ICollection<TableModel.Organization>> SearchAsync(string name, string userEmail,
+                bool? paid, int skip, int take)
```
Comment (cscharf) [inline src/Core/Repositories/EntityFramework/OrganizationRepository.cs:57]: odd spacing issue

File: src/Core/Repositories/EntityFramework/UserRepository.cs [hunk 3]
```diff
@@ -23,7 +23,8 @@ public async Task<TableModel.User> GetByEmailAsync(string email)
             using (var scope = ServiceScopeFactory.CreateScope())
             {
                 var dbContext = GetDatabaseContext(scope);
-                return await GetDbSet(dbContext).FirstOrDefaultAsync(e => e.Email == email);
+                var entity = await GetDbSet(dbContext).FirstOrDefaultAsync(e => e.Email == email);
```
Comment (cscharf) [inline src/Core/Repositories/EntityFramework/UserRepository.cs:26]: Are you considering case-sensitivity of various DB collations of mySQL and/or PostgreSQL to ensure value matches on strings like this are either case-insensitive OR data is normalized on the input to the DB via the provider? Our SQL Server configuration, for example, uses a `_CI` collation so this check would pass regardless of case, but other DBs may not be the same.

File: src/Core/Repositories/SqlServer/Repository.cs [hunk 4]
```diff
@@ -43,16 +43,21 @@ public virtual async Task<T> GetByIdAsync(TId id)
...
 
-        public virtual async Task CreateAsync(T obj)
+        public virtual async Task<T> CreateAsync(T obj)
         {
             obj.SetNewId();
             using (var connection = new SqlConnection(ConnectionString))
             {
+                var parameters = new DynamicParameters();
+                parameters.AddDynamicParams(obj);
+                parameters.Add("Id", obj.Id, direction: ParameterDirection.InputOutput);
```
Comment (cscharf) [inline src/Core/Repositories/SqlServer/Repository.cs:53]: What happens if the object, `T` has `Id` defined already here?

File: test/Core.Test/Repositories/EntityFramework/OrganizationRepositoryTests.cs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+using Bit.Core.Test.Repositories.EntityFramework.EqualityComparers;
+using Bit.Core.Test.AutoFixture.OrganizationFixtures;
+
+namespace Bit.Core.Test.Repositories.EntityFramework
+{
+    public class OrganizationRepositoryTests
+    {
+        [Theory, EfOrganizationAutoData]
+        public async void CreateAsync_Works_DataMatches(
+                Organization organization,
```
Comment (cscharf) [inline test/Core.Test/Repositories/EntityFramework/OrganizationRepositoryTests.cs:19]: These are indented too much here.

File: test/Core.Test/Repositories/EntityFramework/OrganizationRepositoryTests.cs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+            var sqlUser = await sqlOrganizationRepo.CreateAsync(organization);
+            savedOrganizations.Add(await sqlOrganizationRepo.GetByIdAsync(sqlUser.Id));
+
+            var distinctItems = savedOrganizations.Distinct(equalityComparer);
+            Assert.True(!distinctItems.Skip(1).Any());
+        }
+
+        [Theory, EfOrganizationAutoData]
+        public async void ReplaceAsync_Works_DataMatches(Organization postOrganization,
+                Organization replaceOrganization, SqlRepo.OrganizationRepository sqlOrganizationRepo,
```
Comment (cscharf) [inline test/Core.Test/Repositories/EntityFramework/OrganizationRepositoryTests.cs:42]: same

**Review by cscharf — COMMENTED** [review]
Just some quick feedback after initial review.

File: src/Core/Enums/SupportedDatabaseProviders.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+namespace Bit.Core.Enums
+{
+    public enum SupportedDatabaseProviders 
+    {
+        SqlServer,
+        MySql,
+        Postgres
```
[the remark this answers is not in the document]
Comment (addisonbeck) [reply src/Core/Enums/SupportedDatabaseProviders.cs]: Fixed

File: src/Core/Models/Table/SsoConfig.cs [hunk 8]
```diff
@@ -13,7 +13,8 @@ public class SsoConfig : ITableObject<long>
         
         public void SetNewId()
         {
-            // nothing - int will be auto-populated
+            // int will be auto-populated
+            Id = 0;
```
[the remark this answers is not in the document]
Comment (addisonbeck) [reply src/Core/Models/Table/SsoConfig.cs:17]: Doing it that way would still require a blank interface implementation of `SetNewId()` to match the generic constraints that are in place. If this method has to exist, I'd say it is better to let it do something.

File: src/Core/Repositories/EntityFramework/OrganizationRepository.cs (hunk 2 above)
Comment (addisonbeck) [reply src/Core/Repositories/EntityFramework/OrganizationRepository.cs]: Fixed, here and elsewhere

File: src/Core/Repositories/EntityFramework/UserRepository.cs (hunk 3 above)
Comment (addisonbeck) [reply src/Core/Repositories/EntityFramework/UserRepository.cs:26]: This was a fun one to address.

No, I was not considering collation, and Postgres has some hoops to jump through in that regard.
1. There is no option for a database-level collation rule to be applied.
2. `LIKE` clauses don't work for case-insensitive collations.

To handle these I:
1. Created a case-insensitive collation [here](https://github.com/bitwarden/server/pull/1265/files#diff-5b19f4a346a42e5bf955ea7b5ebb0fea518559757ac96f08636ee88d298700edR80-R86) and applied it to any string columns that are being used to look up entities.
2. For methods that use `LIKE` searches on fields that would typically use case-insensitive collations I added some provider-specific logic (😔 ) for postgres to use the default collation & their `ILIKE` function that is case-insensitive.

File: src/Core/Repositories/SqlServer/Repository.cs (hunk 4 above)
Comment (addisonbeck) [reply src/Core/Repositories/SqlServer/Repository.cs:53]: Nothing. I couldn't find any documentation on this in Dapper, but my own debugging and the tests for entities that actually get IDs created by `SetNewId()` like User and Organization show that the flow is:
1. obj.Id gets set by `SetNewId()`
2. A template for parameters is created by `AddDynamicParams(obj)` that includes `obj.Id`
3. `parameters.Add(id)` creates a set-in-stone parameter, not a template like above, with the same Id value
4. Id is returned from the proc execution like normal.

File: test/Core.Test/Repositories/EntityFramework/OrganizationRepositoryTests.cs (hunk 5 above)
Comment (addisonbeck) [reply test/Core.Test/Repositories/EntityFramework/OrganizationRepositoryTests.cs]: Fixed here and elsewhere

File: test/Core.Test/Repositories/EntityFramework/OrganizationRepositoryTests.cs (hunk 6 above)
Comment (addisonbeck) [reply test/Core.Test/Repositories/EntityFramework/OrganizationRepositoryTests.cs]: Fixed here and elsewhere

**Review by cscharf — COMMENTED** [review]
What’s up with this empty file? `test/Core.Test/Utilities/GlobalSettingsFactory.cs`

I think otherwise looks good to go, although let's try to figure out how to get those unit tests, at least for now, to not run for CI.

**Comment by addisonbeck** [comment]
Empty file is no more. Unit tests are ignored in CI, but will still run using `dotnet test` locally.

**Review by cscharf — APPROVED** [review]

**Pull request merged**